lib/YuiRestClient.pm

```diff
@@ -17,110 +17,129 @@ use constant API_VERSION => 'v1';
 
 use testapi;
 use utils qw(enter_cmd_slow type_line_svirt save_svirt_pty zypper_call);
-use Utils::Backends qw(is_pvm is_ipmi);
+use Utils::Backends;
 use YuiRestClient::App;
 use YuiRestClient::Wait;
+use Utils::Architectures 'is_s390x';
 
-our $interval = 1;
-our $timeout  = 10;
-our $app;
+my $app;
+my $port;
+my $host;
 
-sub set_interval {
-    $interval = shift;
+sub get_app {
+    my (%args) = @_;
+    $app = init_app(%args) unless $app;
+    return $app;
 }
 
-sub set_timeout {
-    $timeout = shift;
+sub get_host {
+    return $host;
 }
 
-sub set_app {
-    $app = shift;
+sub get_port {
+    return $port;
 }
 
-sub get_app {
-    return $app;
+sub set_interval {
+    my ($interval) = @_;
+    $app->get_widget_controller()->set_interval($interval);
 }
 
-sub get_yui_params_string {
-    my $port = get_required_var('YUI_PORT');
-    return "YUI_HTTP_PORT=$port YUI_HTTP_REMOTE=1 YUI_REUSE_PORT=1";
+sub set_timeout {
+    my ($timeout) = @_;
+    $app->get_widget_controller()->set_timeout($timeout);
 }
 
-sub connect_to_app {
-    my $port = get_required_var('YUI_PORT');
-    my $host = get_required_var('YUI_SERVER');
-    die "Cannot set libyui REST API server" unless $host;
-    record_info('PORT',   "Used port for libyui: $port");
-    record_info('SERVER', "Connecting to: $host");
-    my $app = YuiRestClient::App->new({
+sub init_app {
+    my (%args)       = @_;
+    my $timeout      = $args{timeout}  || 10;
+    my $interval     = $args{interval} || 1;
+    my $installation = $args{installation};
+    $port = init_port();
+    $host = init_host($installation);
+
+    $app = YuiRestClient::App->new({
             port        => $port,
             host        => $host,
-            api_version => API_VERSION});
-    # As we start installer, REST API is not instantly available
-    $app->connect(timeout => 500, interval => 10);
-    set_app($app);
+            api_version => API_VERSION,
+            timeout     => $timeout,
+            interval    => $interval});
 }
 
-sub connect_to_running_app {
-    return get_app()->connect(timeout => 30, interval => 2);
+sub init_port {
+    return $port if defined $port;
```

Comment:
it looks like you don't need global variables $port and $host, we just need to check the port in app object. makes sense?

Reply:
I was also thinking like that, but then realized that sometimes we need to know port before we initialize App. For example, we need to specify port in boot parameters and then use it after libyui is set up. At the same time, we cannot initialize App on this stage, because IP of the machine may be unknown (e.g. xen or s390x).

$host I left here as global variable just to be consistent, that both variables are come from YuiRestClient.